src/thread.rs

```diff
@@ -63,27 +62,38 @@ impl Pool {
 
     /// Spawn a sub-thread which is joined at the same time as all the
     /// remaining threads.
-    pub fn spawn<N, F, T>(&mut self, name: N, f: F) -> Result<(), std::io::Error>
+    pub fn spawn<N, F>(&mut self, name: N, f: F) -> Result<(), std::io::Error>
     where
-        N: Into<String> + Copy,
-        F: FnOnce() -> T,
+        N: Into<String>,
+        F: FnOnce() -> () + std::panic::UnwindSafe,
         F: Send + 'static,
-        T: Send + 'static,
     {
-        let builder = thread::Builder::new().name(name.into());
+        let name = name.into();
+        let builder = thread::Builder::new().name(name.clone());
 
         let tx = self.tx.clone();
         let handle = builder.spawn(move || {
-            let certificate = DeathCertificate { tx };
-
-            f();
-            drop(certificate);
+            let thread_id = thread::current().id();
+            match std::panic::catch_unwind(|| f()) {
+                Ok(()) => {}
+                Err(panic) => tx
+                    .send(Dead {
+                        thread_id,
+                        cause: Cause::Paniced(panic),
+                    })
+                    .expect("failed to send thread shut-down message!"),
+            }
+            tx.send(Dead {
+                thread_id,
+                cause: Cause::Natural,
+            })
```

Comment:
This must be in the `Ok` above, or two messages are sent

Reply:
I changed it to return only the cause in the match itself and send one message immediately after.